[WIP] Linear algebra

dithillobothrium opened the pull request:
I started to transfer the code related to linear algebra from CP2K to DBCSR. The first part is dbcsr_vector.F . It is not finished, so I created the PR for a discussion.

dithillobothrium commented:
So, now it is compiling. Should test now.

dithillobothrium commented:
If I can use `dbcsr_api` in `arnoldi_api`, how can I use internal dbcsr functions with dbcsr_api?
I cannot pass   `matrix % prv` in dbcsr functions because it is private.

alazzaro commented:
I'm sorry, I don't fully understand what you want to achieve... 
Let me start with a question: do you want to import arnoldi in DBCSR as it is? If this is the case, then I don't see any point in accessing internal DBCSR functionalities, which are not exported by the API. Arnoldi (and any other CP2K code) is already using the dbcsr_api. There is no need to access to any internal DBCSR function, otherwise the entire concept of API would be pointless...

Of course, arnoldi can use MPI from DBCSR (or other existing functionalities which are not DBCSR specific).

Things are different if you want to add new stuff to arnoldi (or any other part of LA) and you want to use other DBCSR functions for your convenience. In this case we might thing to add to the DBCSR API.
Finally, the third case is if you want to access (read-only) to parameters internal to DBCSR, which are not exported. There is a specific function for that (dbcsr_get_info) and we can extend it. Please, do not use any of

  ! to remove:
   PUBLIC :: dbcsr_setname
   PUBLIC :: dbcsr_get_matrix_type
   PUBLIC :: dbcsr_get_occupation
   PUBLIC :: dbcsr_nblkrows_total
   PUBLIC :: dbcsr_nblkcols_total
   PUBLIC :: dbcsr_get_num_blocks
   PUBLIC :: dbcsr_get_data_size
   PUBLIC :: dbcsr_has_symmetry
   PUBLIC :: dbcsr_nfullrows_total
   PUBLIC :: dbcsr_nfullcols_total
   PUBLIC :: dbcsr_get_stored_coordinates
   PUBLIC :: dbcsr_valid_index
   PUBLIC :: dbcsr_get_data_type

alazzaro commented:
In any case, it would be useful if you can elaborate what you want to do with an example

dithillobothrium commented:
So, Arnoldi is a part of LA. I decided to move it first because it will be used in tight binding example for DBCSR. Yes, Arnoldi uses `dbcsr_api`, but also `dbcsr_vector`. So, now I decided to wrap `dbcsr_vector` functions  and place this wrapper to `dbcsr_api` (it is only 5 functions). In this case Arnoldi will use only `dbcsr_api`.

oschuett commented:
@dithillobothrium, I once made an attempt to separate the Arnoldi code more cleanly from `dbcsr_type`: https://github.com/oschuett/cp2k_dev/commits/dbcsr_vector

My overall goal was to make `dbcsr_get_data_p` private because it still allows direct access to internal data structures.

In hindsight, I belive it was a design mistake to use `dbcsr_type` as storage vehicle for vectors (my bad).
So, my plan was to first introduce an actual `dbcsr_vector_type` which I could then reimplement without relying on `dbcsr_type`.

Unfortunately, I never manage to finish the project. However, if you are interested I'm happy to share ideas.

alazzaro commented:
+1 for Ole's idea

oschuett commented:
Ok, then I'll try to dust off my changes from back then. It always bother me that I didn't finish this one.

dithillobothrium commented:
Some changes which are temporary and are needed to be discussed:

- I commented out all the appearances of `abnormal_value` in `iterate_matrix.F`. 
- Copied `ifac` and `maxfac` from `mathlib` to `iterate_matrix.F`

Now Arnoldi compiles. It can be tested together with `dbcsr_vector`, because only Arnoldi and iterate matrix uses dbcsr_vector (it seems so)

dithillobothrium commented:
cmake does not work, probably I need to include new files for compilation in cmake.

The question: how to organize API for `iterate_matrix` and `arnoldi_api`? Make a  wrapper in dbcsr/src /like `dbcsr_la_api.F` ?

alazzaro commented:
Question: since dbcsr_vector is only used in arnoldi, would it make sense to add it under arnoldi and make it internal? Is it used in CP2K somewhere?
Concerning the API, yes, dbcsr_la_api.F is fine. Just look at the tensors directory. The LA should have its own unittest (again, like tensors).
Finally, to avoid conflict, I would propose to rename all files with dbcsr_ prefix.

alazzaro commented:
> * I commented out all the appearances of `abnormal_value` in `iterate_matrix.F`.

What is this function? You can import under utils/dbcsr_toollib.F

> * Copied `ifac` and `maxfac` from `mathlib` to `iterate_matrix.F`
> 
> Now Arnoldi compiles. It can be tested together with `dbcsr_vector`, because only Arnoldi and iterate matrix uses dbcsr_vector (it seems so)

Same here, you can import this function in utils/dbcsr_toollib.F

dithillobothrium commented:
> Question: since dbcsr_vector is only used in arnoldi, would it make sense to add it under arnoldi and make it internal? Is it used in CP2K somewhere?

I think it should be separate because often users need matrix and vectors at the same time without Arnoldi or any iterative method. For example as in my spin dynamics code.

> Concerning the API, yes, dbcsr_la_api.F is fine. Just look at the tensors directory. The LA should have its own unittest (again, like tensors).

OK

> Finally, to avoid conflict, I would propose to rename all files with dbcsr_ prefix.

I will do

dithillobothrium commented on src/ops/dbcsr_vector.F:
```diff
@@ -0,0 +1,1018 @@
+   use dbcsr_block_access,              only: dbcsr_reserve_all_blocks
+
+
+   USE dbcsr_api,                       ONLY: &
```
Currently dbcsr_vector does not use dbcsr_api

dithillobothrium commented:
So, I fixed new dbcsr_vector, added mat x vec test.  I solved the leaking communicators but it crashes at the end with 

`MPI_ABORT was invoked on rank 0 in communicator MPI_COMM_WORLD
with errorcode 1.`

That is unclear up to now.

dithillobothrium commented:
Now cp2k test `tests/QS/regtest-almo-strong/atomic-water.inp` fails with

 *******************************************************************************
 *   ___                                                                       *
 *  /   \                                                                      *
 * [ABORT]                                                                     *
 *  \___/     MPI error 5 in mpi_alltoallv @ mp_alltoall_d11v : MPI_ERR_COMM:  *
 *    |                            invalid communicator                        *
 *  O/|                                                                        *
 * /| |                                                                        *
 * / \                                                    dbcsr_mpiwrap.F:1122 *
 *******************************************************************************


 ===== Routine Calling Stack ===== 

           11 mp_alltoall_d11v
           10 hybrid_alltoall_any
            9 dbcsr_desymmetrize_deep
            8 matrix_qs_to_almo
            7 almo_scf_init_ao_overlap
            6 almo_scf_init
            5 almo_entry_scf
            4 qs_energies
            3 qs_forces
            2 qs_mol_dyn_low
            1 CP2K


Is it related to the new stuff? I did not touch these functions (all2all) at all

alazzaro commented:
It seems an MPI distribution problem. Is the error coming from the CI? Could you reproduce with a different MPI distribution? Also, the DBCSR CI is failing, could you fix it?

dithillobothrium commented:
@oschuett , 

>The clean solution would be implement a stand-alone dbcsr_vector_type which simply uses a plain array instead for storage.

Yes, I think it could be done, but maybe later. After the vector will be included in dbcsr and cp2k, we can start to refactor. Actually it is already working, but some tests are needed. 

Now I have plans to make an example of using the iterate_matrix.F algorithms. In parallel I think, I will be able to discuss about the refactor. 

> The most tricky part about the code is how it uses two consecutive broadcasts to replicate and transpose the vector. If you are interested, I'm happy to answer your questions.

I had questions regarding this part, because it is not clear how it works

dithillobothrium commented:
it fails again

dithillobothrium commented:
so, now it does not pass coverage tests. It seems that it is not only my code
